include/deal.II/fe/fe_values_base.h

```diff
@@ -1628,9 +1635,16 @@ class FEValuesBase : public Subscriptor
                                 Vector<IndexSet::value_type> &out) const;
 
   private:
-    std::optional<typename Triangulation<dim, spacedim>::cell_iterator> cell;
-    const DoFHandler<dim, spacedim> *dof_handler;
-    bool                             level_dof_access;
+    /**
+     * The cell in question, if one has been assigned to this object. The
+     * concrete data type can either be a Triangulation cell iterator, a
+     * DoFHandler cell iterator, or a DoFHandler level cell iterator.
+     */
+    std::optional<
+      std::variant<typename Triangulation<dim, spacedim>::cell_iterator,
+                   typename DoFHandler<dim, spacedim>::cell_iterator,
+                   typename DoFHandler<dim, spacedim>::level_cell_iterator>>
+      cell;
```

Comment:
This is the key part of the patch. We really do store a variant already, we just break the data stored into the base object plus potentially the members of a derived class. Using the `std::variant` is, in my view, a clear win.